File: Data Structures & Algorithms/count-connected-components/submission-18.js
class DSU {
    /**
     * @param {number} n
     */
    constructor(n) {
        this.parent = Array.from({ length: n }, (_, i) => i);
        this.rank = Array(n).fill(1);
    }

    /**
     * @param {number} node
     * @return {number}
     */
    find(node) {
        let cur = node;
        while (cur !== this.parent[cur]) {
            this.parent[cur] = this.parent[this.parent[cur]];
            cur = this.parent[cur];
        }
        return cur;
    }

    /**
     * @param {number} u
     * @param {number} v
     * @return {boolean}
     */
    union(u, v) {
        let pu = this.find(u);
        let pv = this.find(v);
        if (pu === pv) {
            return false;
        }
        if (this.rank[pv] > this.rank[pu]) {
            [pu, pv] = [pv, pu];
        }
        this.parent[pv] = pu;
        this.rank[pu] += this.rank[pv];
        return true;
    }
}

class Solution {
    /**
     * @param {number} n
     * @param {number[][]} edges
     * @returns {number}
     */
    countComponents(n, edges) {
        const dsu = new DSU(n);
        let res = n;
        for (const [u, v] of edges) {
            if (dsu.union(u, v)) {
                res--;
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/valid-parenthesis-string/submission-18.cs
public class Solution {
    public bool CheckValidString(string s) {
        int n = s.Length;
        bool[,] dp = new bool[n + 1, n + 1];
        dp[n, 0] = true;

        for (int i = n - 1; i >= 0; i--) {
            for (int open = 0; open < n; open++) {
                bool res = false;
                if (s[i] == '*') {
                    res |= dp[i + 1, open + 1];
                    if (open > 0) res |= dp[i + 1, open - 1];
                    res |= dp[i + 1, open];
                } else {
                    if (s[i] == '(') {
                        res |= dp[i + 1, open + 1];
                    } else if (open > 0) {
                        res |= dp[i + 1, open - 1];
                    }
                }
                dp[i, open] = res;
            }
        }
        return dp[0, 0];
    }
}